### DISCUSSION TITLE
How to interop with opentk via RenderTargetBitmap

### DISCUSSION THREAD

**Discussion by BreyerW** [AvaloniaUI/Avalonia#5473] [Q&A] [answered]
Im thinking on porting hobby project to avalonia but already use opentk rather extensively so i would like to avoid using avalonia's opengl surface (using avalonia api for initialization is fine by me but actual 
rendering should be done via opentk. I can use low lvl api like `getprocaddress` and others if thats necessary). Is this possible and if yes what api i should use? I dont want to embed native window if possible since i might need rendering avalonia controls over window that gets image from rendering. Im using last opentk 3 version since opentk 4 killed easy way to wire opentk to any windowing library (they are halfway there funnily enough with `IBindingsContext` but theres no way to load custom graphic context other than glfw one and i guess thats the harder half)

Another challenge would be to avoid copying data. I saw very simple example with `WriteableBitmap` but that copies data if im not mistaken. Digging through docu i saw `RenderTargetBitmap` that sounds promising but couldnt see examples how to use it and if it allow avoiding copying?

Extra point: 

Im thinking would it be possible to have two threads one for avalonia and other for opentk and interop rendering via `RenderTargetBitmap` or similar primitive that could be displayed in window?

Thanks in advance for any pointers

**Comment by kekekeks** [comment, marked as answer]
See https://github.com/AvaloniaUI/Avalonia/blob/6332dfca96430e45cf61a82d28d5765358175d1a/samples/ControlCatalog/Pages/OpenGlPage.xaml.cs#L1 for OpenGL rendering example. You can get glGetProcAddress from provided context.

Also take a look at https://github.com/AvaloniaUI/Avalonia/blob/6332dfca96430e45cf61a82d28d5765358175d1a/src/Avalonia.OpenGL/Controls/OpenGlControlBase.cs itself since it only uses our public API surface.

There are two limitations though:
1) on Windows you need to add `.With(new Win32PlatformOptions {UseWgl = true}`
2) there is some bug with DPI support that I haven't investigated yet

**Reply by BreyerW** [reply]
Alright thanks for quick answer. Is renderTargetBitmap completely useless in this scenario? Ideally would like to decouple general rendering from avalonia rendering. This potentially would allow for naive multithreading and also i will need multiple windows with different view (like blender view with 4 split. Each split for different dimension+1 split for normal 3d viewing) or preview subwindow for shader as example. 

One of ways to achieve that would be to have 1 render loop with framebuffer for each window/subwindow and bind these framebuffers to some sort of image/bitmap source on avalonia side (note not necessarily on different thread but that would be ideal)

If that is unreasonable/impossible then i guess i will try  what you showed

**Reply by kekekeks** [reply]
> Ideally would like to decouple general rendering from avalonia rendering.

Create your OpenGL context manually and use `OpenGlBitmap` then. That will allow you to render on a separate thread.

**Reply by BreyerW** [reply]
Ohh thats perfect. Thanks!
